Multiarch image support

Prajyot-Parab opened the pull request:
- added multi-arch image support

closes #1645

nrb commented on hack/get-restic-ppc64le.sh:
```diff
@@ -0,0 +1,30 @@
+    exit 1
+fi
+
+wget --quiet https://oplab9.parqtec.unicamp.br/pub/ppc64el/restic/restic-${RESTIC_VERSION}
```
Who maintains this restic build?

Prajyot-Parab replied:
Restic builds are maintained by my team and hosted on that link. Also, we are already in the process to get it supported on power with Restic community.

Prajyot-Parab replied:
Yes our plan is to switch to GH release once it's up
CC - @rpsene

nrb commented on Makefile:
```diff
@@ -33,12 +33,16 @@ VERSION ?= master
 
 CLI_PLATFORMS := linux-amd64 linux-arm linux-arm64 darwin-amd64 windows-amd64 linux-ppc64le
-CONTAINER_PLATFORMS := linux-amd64 linux-arm linux-arm64 linux-ppc64le
+CONTAINER_PLATFORMS := linux-amd64 linux-ppc64le #linux-arm linux-arm64
```
Why are arm platforms removed?

Prajyot-Parab replied:
arm is commented out of CONTAINER_PLATFORMS as there are no images being maintained for them at the moment. The cross-compile binaries will be built for those architectures as they are part of CLI_PLATFORMS. Also, the current workflow is building images explicitly for linux-amd64 so the arm platforms are not utilized for image building for now.

nrb commented on Makefile:
```diff
@@ -147,12 +167,23 @@ ifeq ($(TAG_LATEST), true)
+	docker manifest push $(MULTIARCH_IMAGE):latest
+endif
+
+#all-containers:
```
These `all-containers` and `all-push` targets should be removed if we don't need them.

Prajyot-Parab replied:
`all-containers` - will build images for each of the supported arch `CONTAINER_PLATFORMS := linux-amd64 linux-ppc64le`

`all-push` - will push all of this arch images

This will be followed by `all-manifest`
which will create a manifest out of this arch images and push it.

nrb replied:
I'm specifically talking about the commented out blocks here; they should be removed as the definition changed.

Prajyot-Parab replied:
Done

Prajyot-Parab commented:
> Several inline comments. Also, can you provide documentation for how this should be invoked?
> 
> I tried the following on linux amd64:
> 
> ```shell
> x1c in /home/nrb/go/src/github.com/heptio/velero (git) Prajyot-Parab-multiarch-image-support
> % make all-manifest
> make[1]: Entering directory '/home/nrb/go/src/github.com/heptio/velero'
> no such manifest: gcr.io/heptio-images/velero-amd64:master
> make[1]: *** [Makefile:215: .manifest-gcr.io/heptio-images/velero] Error 1
> make[1]: Leaving directory '/home/nrb/go/src/github.com/heptio/velero'
> make: *** [Makefile:102: all-manifest] Error 2
> 
> x1c in /home/nrb/go/src/github.com/heptio/velero (git) Prajyot-Parab-multiarch-image-support
> % make all-manifest
> make[1]: Entering directory '/home/nrb/go/src/github.com/heptio/velero'
> no such manifest: gcr.io/heptio-images/velero-amd64:master
> make[1]: *** [Makefile:215: .manifest-gcr.io/heptio-images/velero] Error 1
> make[1]: Leaving directory '/home/nrb/go/src/github.com/heptio/velero'
> make: *** [Makefile:102: all-manifest] Error 2
> 
> x1c in /home/nrb/go/src/github.com/heptio/velero (git) Prajyot-Parab-multiarch-image-support
> % make manifest
> no such manifest: gcr.io/heptio-images/velero-amd64:master
> make: *** [Makefile:215: .manifest-gcr.io/heptio-images/velero] Error 1
> ```

The error is observed as the image tagging is changed and now it's looking for images like
`gcr.io/heptio-images/velero-amd64:master`
`gcr.io/heptio-images/velero-ppc64le:master`
(once these images have been pushed you won't encounter above issue in future)
as now the final manifest will be tagged as `gcr.io/heptio-images/velero:master`


Try this workflow

`make all-build all-containers all-push all-manifest`

1. Build binaries for all the platforms part of ( CLI_PLATFORMS list)
`Binaries will be created for : linux-amd64 linux-arm linux-arm64 darwin-amd64 windows-amd64 linux-ppc64le `

2. Build images for all the platforms part of ( CONTAINER_PLATFORMS list)
`Images will be created for : linux-amd64 linux-ppc64le`
Example:
`gcr.io/heptio-images/velero-amd64:master`
`gcr.io/heptio-images/velero-ppc64le:master`

3. Push all the images part of ( CONTAINER_PLATFORMS list)
`Push images for :  linux-amd64 linux-ppc64le` i.e above images

4. Create and push manifest 
Example of manifest:
`gcr.io/heptio-images/velero:master`

carlisia commented:
The commented out code should be removed if not used now. We can always recover it later by going thru the git history if we want to bring any of it back.

Prajyot-Parab commented:
> The commented out code should be removed if not used now. We can always recover it later by going thru the git history if we want to bring any of it back.

Agreed

Prajyot-Parab commented on Makefile:
```diff
@@ -70,18 +80,28 @@ all:
 all-build: $(addprefix build-, $(CLI_PLATFORMS))
 
-#all-container: $(addprefix container-, $(CONTAINER_PLATFORMS))
+all-container: $(addprefix container-, $(CONTAINER_PLATFORMS))
```
Done

Prajyot-Parab commented on Makefile:
```diff
@@ -70,18 +80,28 @@ all:
+all-push: $(addprefix push-, $(CONTAINER_PLATFORMS))
 
-#all-push: $(addprefix push-, $(CONTAINER_PLATFORMS))
+all-manifest:
```
Done

Prajyot-Parab commented on changelogs/unreleased/1768-prajyot-parab:
```diff
@@ -0,0 +1 @@
+Added support for multi-arch images
```
Done

Prajyot-Parab commented on Makefile:
```diff
@@ -48,6 +52,8 @@ GOARCH = $(word 2, $(platform_temp))
 # Set default base image dynamically for each arch
 ifeq ($(GOARCH),amd64)
 		DOCKERFILE ?= Dockerfile-$(BIN)
+local-arch:
```
`local-arch` - For Power, this target would run a script that would download the Restic binary

As we are creating arch images (except amd64) without using dedicated arch nodes this will be needed (also in case of downloading the binary from Restic GH releases in future), this can be used in the future while downloading Restic binary for arm platforms.

skriss replied:
I see that L66-68 are doing the restic download, however still unclear on the purpose of L55-56

Prajyot-Parab replied:
L55-56 is added to avoid the workflow failure due to local-arch target missing while building for amd64 as in this case, it will be downloaded via Dockerfile. L155 will always trigger local-arch target and L55-56 will help us avoid any failure.

Do let me know if any further clarification is needed on this.

clnperez commented:
@nrb -- where do you think makes the most sense for the `all-manifests` target to be called?

nrb commented:
@clnperez I'm not sure I understand your question entirely; do you mean where will it be invoked by CI/CD?

If that's the case, I'm currently working on getting our container images published via a deploy script on Travis CI, which is likely where it would go.

clnperez commented:
@nrb -- yep. wasn't sure if that was something you wanted @Prajyot-Parab to add to the PR or if it was something you needed to add

nrb commented:
@clnperez Cool - I think to prevent your team from constantly changing this, we could merge this before the Travis work. The biggest hang up I think we have is that the Power restic binary is coming from a 3rd party instead of the original developer's repo.

I recognize that the restic repo has slowed in merging changes and making releases, so I'll see if we can talk to our legal team to see what they'd advise.

rpsene commented:
@nrb let me know if you want to have some validation of the content we have at Unicamp, I can create a new build mechanism while we do not get the merge approved on the restic side. We can audit anything you want to ensure the package integrity.

nrb commented:
> @nrb -- where do you think makes the most sense for the `all-manifests` target to be called?

I have a more definitive answer on this now - it would be at https://github.com/heptio/velero/blob/master/hack/gcr-push.sh#L75

onedr0p commented:
lgtm!

carlisia commented:
@nrb / @skriss does this have to be on hold pending something else, or is it good to merge?

nrb commented:
@carlisia We're waiting for https://github.com/restic/restic/pull/2342 to be merged. Honestly, the biggest thing I want right now is to have an official restic build so we don't forget to change from the [unicamp URL](https://github.com/vmware-tanzu/velero/pull/1768/files#diff-ecd719c0f4b520cae14dfb54edec7bffR27) featured in this PR.

nrb commented:
Now that https://github.com/restic/restic/pull/2342 is merged, I feel better about accepting this. We'll still need a follow up action to change the URL, but we don't know how long it will be before a new restic release, and I also don't want to keep this in limbo forever.

Plus, as we move towards GitHub Actions, this will bitrot faster.

I'll give this another look this week and see if there's anything else that needs to be addressed, but I'm going to remove my objection over the URL for now and just say that we'll create an issue to track and update it.

@skriss do you want to look over this again as well before the holiday break?

skriss commented:
yeah I'll take a fresh look at this today.

nrb commented:
@Prajyot-Parab Could you provide workflow described in [this comment](https://github.com/vmware-tanzu/velero/pull/1768#issuecomment-522851088) as documentation for building the manifest in this PR?

Prajyot-Parab commented:
> @Prajyot-Parab Could you provide workflow described in [this comment](https://github.com/vmware-tanzu/velero/pull/1768#issuecomment-522851088) as documentation for building the manifest in this PR?

@nrb Please take a look at this [Manifest doc](https://github.com/Prajyot-Parab/HowTo/wiki/Velero-Manifest) and let me know about your thoughts on the same. I will add it to PR (after your review/approval)

skriss commented on Makefile:
```diff
@@ -33,12 +33,16 @@ VERSION ?= master
 TAG_LATEST ?= false
 
+# The version of restic binary to be downloaded for power architecture
+RESTIC_VERSION ?= 0.9.4
```
let's bump this to v0.9.5 to keep it consistent with our other images (can you confirm you have that version in your repo?)

Prajyot-Parab replied:
@skriss Yes we have that version in our repo

Prajyot-Parab replied:
Done

skriss commented on Makefile:
```diff
@@ -57,31 +63,42 @@ endif
                 DOCKERFILE ?= Dockerfile-$(BIN)-ppc64le
+local-arch:
+	RESTIC_VERSION=$(RESTIC_VERSION) \
+        ./hack/get-restic-ppc64le.sh
```
is there a reason we can't do the `restic` binary download within the container image build, similarly to how we're doing it for the non-Power image builds?

Prajyot-Parab replied:
1. L55-56 is added to avoid the workflow failure due to local-arch target missing while building for amd64 as in this case, it will be downloaded via Dockerfile. L155 will always trigger local-arch target and L55-56 will help us avoid any failure.

2. We cant do restic binary download within the container image as this will run the workflow on amd64 machine and it will cause exec error due to arch issues if we do it via container as the Dockerfile for power will use a ppc64le arch image.

Do let me know if any further clarification is needed on this.

skriss commented on hack/docker-push.sh:
```diff
@@ -80,4 +80,4 @@ unset GIT_HTTP_USER_AGENT
 echo "Building and pushing container images."
 
-VERSION="$VERSION" TAG_LATEST="$TAG_LATEST" make all-containers all-push
+VERSION="$VERSION" TAG_LATEST="$TAG_LATEST" make all-build all-containers all-push all-manifests
```
do we need to add `all-build` here? I think the `all-containers` prerequisite chain is sufficient for getting the binaries built prior to building the containers

skriss replied:
should be fine as-is.

skriss commented on Makefile:
```diff
@@ -33,12 +33,16 @@ VERSION ?= master
 ### These variables should not need tweaking.
 ###
 
 CLI_PLATFORMS := linux-amd64 linux-arm linux-arm64 darwin-amd64 windows-amd64 linux-ppc64le
```
let's move L43-45 up to starting at L39 (above the "should not need tweaking" comment), and change them all to `?=` instead of `:=`. This is specifically necessary for `MANIFEST_PLATFORMS`, because when building a container locally for development purposes I want to override the value of this variable so I don't have to build a Power image just to push my amd64 dev image.

FYI, this is the command I'm running for dev purposes, that should allow me to build&push *just* an amd64 image:

```
REGISTRY=steveheptio VERSION=foo MANIFEST_PLATFORMS=amd64 make container push manifest
```

skriss commented on Makefile:
```diff
@@ -161,6 +173,20 @@ endif
 
+manifest: .manifest-$(MULTIARCH_IMAGE) manifest-name
+.manifest-$(MULTIARCH_IMAGE):
+	@docker manifest create $(MULTIARCH_IMAGE):$(VERSION) \
```
I'm seeing an issue here where if I try to build/push the same tag repeatedly, I get the following error:

```
refusing to amend an existing manifest list with no --amend flag
```

Unfortunately, adding the `--amend` flag doesn't quite give me the behavior I'm looking for, since it adds to, rather than replaces, the manifest list (e.g. if on build 1 I do amd64 only, then on build 2 I do ppc64le only, my manifest after build 2 contains both).

Seems the only way to delete a manifest list is by manually deleting from `~/.docker/manifests` -- we may need to do that prior to creating them, but want to get others thoughts here before going down that path.

Prajyot-Parab replied:
@clnperez your thoughts on this one (possible workaround)?

clnperez replied:
when you do the push, make sure you use the `--purge` flag to wipe out the old bits that are stored locally. hopefully that works for you!

skriss replied:
ah, that works and seems reasonable, thanks @clnperez!

Prajyot-Parab replied:
@skriss Do let me know if this addresses your concern

Prajyot-Parab replied:
> @skriss Do let me know if this addresses your concern

Noted.

skriss replied:
@Prajyot-Parab please add the `--purge` flag to the `docker manifest push` commands in the Makefile, thanks!

Prajyot-Parab replied:
@skriss Should I go ahead and do these changes ?
`@docker manifest push --purge $(MULTIARCH_IMAGE):$(VERSION)`
`docker manifest push --purge $(MULTIARCH_IMAGE):latest`

nrb commented:
Thanks @Prajyot-Parab and @clnperez for being patient with us and addressing Steve's comments on the Makefile changes.

Also, to be transparent, I'm in the process of moving from Travis CI to GitHub Actions. (#2104) However, since our team is going to be out from Dec 23 2019 to Jan 6 2020, I'm not going to push forward on moving container image builds from Travis to GitHub Actions.

That shouldn't affect this PR, as so far there's nothing touching travis files right as we can simply invoke the appropriate make targets in the GitHub Action files, but I wanted to be sure that everyone was aware that the CI runner work was happening in parallel.

rpsene commented:
> Thanks @Prajyot-Parab and @clnperez for being patient with us and addressing Steve's comments on the Makefile changes.
> 
> Also, to be transparent, I'm in the process of moving from Travis CI to GitHub Actions. (#2104) However, since our team is going to be out from Dec 23 2019 to Jan 6 2020, I'm not going to push forward on moving container image builds from Travis to GitHub Actions.
> 
> That shouldn't affect this PR, as so far there's nothing touching travis files right and we can simply invoke the appropriate make targets in the GitHub Action files, but I wanted to be sure that everyone was aware that the CI runner work was happening in parallel.

I would like to state the GH Actions still does not work on Power, how could we get builds/pr validated on Power? Can we keep Travis in addition to GH Actions?

nrb commented:
> I would like to state the GH Actions still does not work on Power, how could we get builds/pr validated on Power? Can we keep Travis in addition to GH Actions?

Is there something I'm missing in this PR that makes the jobs on Travis run on Power? The Velero builds are happening via Go cross-compilation, and the restic builds are downloaded, correct?

Prajyot-Parab commented:
> > I would like to state the GH Actions still does not work on Power, how could we get builds/pr validated on Power? Can we keep Travis in addition to GH Actions?
> 
> Is there something I'm missing in this PR that makes the jobs on Travis run on Power? The Velero builds are happening via Go cross-compilation, and the restic builds are downloaded, correct?

@nrb Yes Velero builds are cross-compiled, and the restic builds are downloaded. I think @rpsene may be referring to having builds in future for Power (if needed since Travis provides Power node). @rpsene is my understanding inline with yours ?

skriss commented on Makefile:
```diff
@@ -161,6 +173,20 @@ endif
 
+manifest: .manifest-$(MULTIARCH_IMAGE) manifest-name
+.manifest-$(MULTIARCH_IMAGE):
+	@docker manifest create $(MULTIARCH_IMAGE):$(VERSION) \
```
Looks like we need to set `DOCKER_CLI_EXPERIMENTAL=enabled` for any `docker manifest` subcommands to be accessible. I think it makes sense to do it inline here, each place the `docker manifest` cmd is used, rather than setting it more globally for the build.

Prajyot-Parab replied:
@skriss 
Something like this `export DOCKER_CLI_EXPERIMENTAL=enabled` is expected before every phase (2 time within and outside if) where manifest commands are used?

skriss replied:
you could change e.g. L178 to `@DOCKER_CLI_EXPERIMENTAL=enabled docker manifest create $(MULTIARCH_IMAGE):$(VERSION) \`, and similarly for subsequent lines

Prajyot-Parab replied:
@skriss 
Okay, you want me to test it out first? or if you already did, I will go ahead and push the changes

skriss replied:
I tried it and it worked properly for me, but additional testing wouldn't be a bad thing :)

Prajyot-Parab replied:
Done

rpsene commented:
> > > I would like to state the GH Actions still does not work on Power, how could we get builds/pr validated on Power? Can we keep Travis in addition to GH Actions?
> > 
> > 
> > Is there something I'm missing in this PR that makes the jobs on Travis run on Power? The Velero builds are happening via Go cross-compilation, and the restic builds are downloaded, correct?
> 
> @nrb Yes Velero builds are cross-compiled, and the restic builds are downloaded. I think @rpsene may be referring to having builds in future for Power (if needed since Travis provides Power node). @rpsene is my understanding inline with yours ?

@Prajyot-Parab yep, you are right! I'm looking for the future, but as we are cross-compiling, we are good to go!

nrb commented:
Thanks for clarifying! We're moving away from Travis because of longevity and support concerns, for what it's worth.

Hopefully GitHub Actions will get broader capabilities, including PowerPC support, in the future but for this project I think the cross-compilation addresses it.

skriss commented:
We also need to update the docs -- https://github.com/vmware-tanzu/velero/blob/master/site/docs/master/build-from-source.md and https://github.com/vmware-tanzu/velero/blob/master/site/docs/v1.2.0/build-from-source.md -- to reflect the `Makefile` changes.

Specifically, under the "Making images and updating Velero", we need to revise the instructions to cover the manifest flow.  At a minimum, we need to indicate that `make push` will push an image named e.g. `velero-amd64:<tag>`, and that the user needs to run `make manifest` to get the `velero:<tag>` manifest list pushed. It would be nice to also cover that the user can run e.g. `MANIFEST_PLATFORMS=amd64 make manifest` to push a manifest that has *just* the amd64 image, if they aren't interested in a `ppc64le` one (or vice versa).

FWIW, the command I use for building and pushing a container image was previously:

```
REGISTRY=steveheptio VERSION=foo make container push
```

and will now be:

```
REGISTRY=steveheptio VERSION=foo MANIFEST_PLATFORMS=amd64 make container push manifest
```

So if you want to include that in the doc as an example all-encompassing command, that would be great too.

nrb commented on Makefile:
```diff
@@ -48,6 +52,8 @@ GOARCH = $(word 2, $(platform_temp))
 # Set default base image dynamically for each arch
```
Could you remove the todo comment on L51?

Prajyot-Parab replied:
Done

Prajyot-Parab commented:
> We also need to update the docs -- https://github.com/vmware-tanzu/velero/blob/master/site/docs/master/build-from-source.md and https://github.com/vmware-tanzu/velero/blob/master/site/docs/v1.2.0/build-from-source.md -- to reflect the `Makefile` changes.
> 
> Specifically, under the "Making images and updating Velero", we need to revise the instructions to cover the manifest flow. At a minimum, we need to indicate that `make push` will push an image named e.g. `velero-amd64:<tag>`, and that the user needs to run `make manifest` to get the `velero:<tag>` manifest list pushed. It would be nice to also cover that the user can run e.g. `MANIFEST_PLATFORMS=amd64 make manifest` to push a manifest that has _just_ the amd64 image, if they aren't interested in a `ppc64le` one (or vice versa).
> 
> FWIW, the command I use for building and pushing a container image was previously:
> 
> ```
> REGISTRY=steveheptio VERSION=foo make container push
> ```
> 
> and will now be:
> 
> ```
> REGISTRY=steveheptio VERSION=foo MANIFEST_PLATFORMS=amd64 make container push manifest
> ```
> 
> So if you want to include that in the doc as an example all-encompassing command, that would be great too.

@skriss @nrb I updated the above-mentioned docs and added them to this PR. Do let me know in case of any issues with the same.

skriss commented:
@Prajyot-Parab thanks for the docs updates, will look shortly.

Just a note on timing - given that most of the Velero maintainers are about to head out of office for the holidays, we're probably going to defer merging this until the new year. From my point of view, the code looks in good shape (I'll take one more look today) so it should be good to go when we pick it up in 2020.

nrb commented:
Will add to my queue for today. From what I remember before break it should be good, but I'll do one more double check just to refresh my memory.

skriss commented:
@Prajyot-Parab just wanted to circle back here and say thanks for your patience, and for seeing this through!
